qiskit/circuit/add_control.py

```diff
@@ -111,21 +110,22 @@ def control(operation: Union[Gate, ControlledGate],
         ctrl_substr = ('{0}' * new_num_ctrl_qubits).format('c')
     new_name = '{0}{1}'.format(ctrl_substr, base_name)
 
-    # define controlled gate instructions
-
-    if (not operation.definition and
-            not isinstance(operation, (CXGate, U3Gate, IGate))):
-        # opaque gate
-        cgate = controlledgate.ControlledGate(
-            new_name,
-            operation.num_qubits + new_num_ctrl_qubits,
-            operation.params,
-            label=label,
-            num_ctrl_qubits=new_num_ctrl_qubits,
-            definition=None,
-            ctrl_state=ctrl_state)
-        cgate.base_gate = base_gate
-        return cgate
+    # detect opaque gate
+    if not operation.definition:
+        # since u3, cx, and i are opaque-like we want to distinguish them here;
+        if operation.__class__ is Gate:
```

Comment:
`if isinstance(operation, Gate)`?
I think what you have currently does not work for an opaque `ControlledGate`.

Reply:
resolved, thanks.